components/post_view/reaction/reaction.jsx

```diff
@@ -10,88 +10,152 @@ import OverlayTrigger from 'components/overlay_trigger';
-         */
         otherUsersCount: PropTypes.number.isRequired,
-
-        /*
-         * Array of reactions by user
-         */
         reactions: PropTypes.arrayOf(PropTypes.object).isRequired,
-
-        /*
-         * True if the user has the permission to add a reaction in this channel
-         */
         canAddReaction: PropTypes.bool.isRequired,
-
-        /*
-         * True if user has the permission to remove his own reactions in this channel
-         */
         canRemoveReaction: PropTypes.bool.isRequired,
-
-        /*
-         * The URL of the emoji image
-         */
         emojiImageUrl: PropTypes.string.isRequired,
-
         actions: PropTypes.shape({
-
-            /*
-             * Function to add a reaction to a post
-             */
             addReaction: PropTypes.func.isRequired,
-
-            /*
-             * Function to get non-loaded profiles by id
-             */
             getMissingProfilesByIds: PropTypes.func.isRequired,
-
-            /*
-             * Function to remove a reaction from a post
-             */
             removeReaction: PropTypes.func.isRequired,
         }),
         sortedUsers: PropTypes.object.isRequired,
     }
 
-    handleAddReaction = (e) => {
-        e.preventDefault();
-        const {actions, post, emojiName} = this.props;
-        actions.addReaction(post.id, emojiName);
+    constructor(props) {
+        super(props);
+
+        const {reactionCount, canRemoveReaction, canAddReaction} = this.props;
+        const {currentUserReacted} = this.props.sortedUsers;
+
+        if (currentUserReacted) {
+            this.state = {
+                userReacted: currentUserReacted,
+                reactedClass: 'Reaction reacted',
+                displayNumber: reactionCount,
+                reactedNumber: reactionCount,
+                unreactedNumber: reactionCount - 1,
+            };
+        } else {
+            this.state = {
+                userReacted: currentUserReacted,
+                reactedClass: 'Reaction unreacted',
+                displayNumber: reactionCount,
+                reactedNumber: reactionCount + 1,
+                unreactedNumber: reactionCount,
+            };
+        }
+
+        if (currentUserReacted && !canRemoveReaction) {
+            this.state = {
+                ...this.state,
+                reactedClass: 'Reaction read-only',
+            };
+        } else if (!currentUserReacted && !canAddReaction) {
+            this.state = {
+                ...this.state,
+                reactedClass: 'Reaction read-only',
+            };
+        }
+
+        this.reactionButtonRef = React.createRef();
+        this.reactionCountRef = React.createRef();
+        this.reactedNumeralRef = React.createRef();
     }
 
-    handleRemoveReaction = (e) => {
-        e.preventDefault();
-        this.props.actions.removeReaction(this.props.post.id, this.props.emojiName);
+    shouldComponentUpdate(nextProps) {
```

Comment:
We shouldn't be using `setState` here. Could this be done in `getDerivedStateFromProps` instead?

Reply:
I went back and forth on this a lot given the "you probably don't need `getDerivedStateFromProps`, but I think we might in this scenario. I had this logic under there originally.

Reply:
Long story short, I figure out how to do this in `getDerived...`. Because it fires frequently, you have to return the derived state when all of the props conditions are true. If you nest the conditions, it gets confusing as the props trickle in sequentially as they update.